keps/sig-auth/541-external-credential-providers/README.md

```diff
@@ -313,20 +337,57 @@ has completed parsing its `kubeconfig`, flags and environment variables).
 This allows the executable to perform different actions based on the current
 cluster (i.e. get a token for a particular cluster).  The `Cluster` struct is
 flexible in that it not only provides all details required to communicate with
-the cluster (hostname and TLS config), but that is also allows arbitrary
+the cluster as one would via a `kubeconfig` (i.e. everything from
+`"k8s.io/client-go/tools/clientcmd/api/v1".Cluster`), but also allows arbitrary
 per-cluster configuration to be passed to the executable via the `config` field.
 This field can contain arbitrary data that is passed to the executable without
 modification.  This allows extra user-defined data (i.e. an OAuth client ID for
 audience scoping) to be passed through the `spec.cluster` field.  The user
-configures this via the `kubeconfig`'s `clusters.[...].cluster.extensions[exec].extension`
-field.  The `exec` named extension is reserved for this purpose.
+configures this via the `kubeconfig`'s `clusters.[...].cluster.extensions[client.authentication.k8s.io/exec].extension`
+field.  The `client.authentication.k8s.io/exec` named extension is reserved for this
+purpose.
+
+To ensure that the `Cluster` struct maintains the same cluster connection
+capabilities as one gets from a `kubeconfig`, the `Cluster` fields (Go and JSON) must
+be named the same as `"k8s.io/client-go/tools/clientcmd/api/v1".Cluster`, with the
+exception of `InsecureSkipVerify`, which has been left out to discourage its use, and
+`CertificateAuthority`, which has been left out since CA data will always be passed
+to the plugin as bytes.
 
 This data is passed to the executable via the `KUBERNETES_EXEC_INFO` environment
```

Comment:
Open question: make exec plugins opt into getting cluster connection info in (maybe giant) envvar or always send?

Reply:
Strawperson proposal: add boolean field to `"k8s.io/client-go/tools/clientcmd/api/v1".ExecConfig` to turn _off_ passing the `Cluster` data to the exec plugin. Following the 90/10 rule here, I feel like this extra data is not going to affect performance 90% of the time, whereas maybe only 10% would have an issue with a too-large env var or a env var performance problem.

[0] https://github.com/kubernetes/kubernetes/blob/4b24dca228d61f4d13dcd57b46465b0df74571f6/staging/src/k8s.io/client-go/tools/clientcmd/api/v1/types.go#L197

Reply:
It's really just the CA data I'm concerned about... the server name, custom exec config, etc, always seem reasonable to pass as inputs. I'd guess a different invocation of the 90/10 rule would say that 90% of the exec plugins won't need to talk to the server they are providing credentials to, so we should omit potentially breaking data by default and let those that need it opt in in their config.

The issue with making it opt-out instead of opt-in is that the owner of the CA bundle and the author of the kubeconfig/exec config aren't always the same, and relying on the user to modify the guts of a kubeconfig file that was likely generated for them isn't good.

Reply:
> 90% of the exec plugins won't need to talk to the server
Fair. Agreed.

> the owner of the CA bundle and the author of the kubeconfig/exec config aren't always the same
Agreed. Ya know, if we assume[0] that the creator of the kubeconfig is knowledgable of how to properly configure the exec plugin, then it is probably safe to assume that that creator should know whether the exec plugin needs cluster info to work, and therefore the creator should know whether it should turn on this flag.

---

So let's assume we add this flag for opting into passing the CA data to the plugin.
1. It seems like we are establishing a new branch for the exec plugins that want to consume this cluster info (which may be 10% of plugins, to your point): if there is no cluster info do XXX, else do YYY. That's probably fine, because in the worst case, the exec plugin could just fail and say that it is missing cluster info.
1. It seems weird to only remove that one CA data field from the `Cluster` struct. Maybe this is what you were already suggesting, but perhaps we should not set any of the `Cluster` info in the `KUBERNETES_EXEC_INFO` at all if someone doesn't turn on this flag? This would discourage plugins from still trying to use the `Cluster` info to connect to a cluster without verifying the server's certificate using the (missing) CA data.
1. How would we indicate to the plugin (or to ourselves) that the `Cluster` data was not set in the env var? I'm assuming we still want to set the env var for forward compatibility (and potentially some backwards compatibility that I didn't know about until I read your comment below). Perhaps if `Cluster.Server` is empty, then we can assume that no cluster info has been provided?

---

[0] Supporting evidence:
1. `gcloud container clusters get-credentials` adds itself as the auth plugin for your kubeconfig.
1. `az aks get-credentials` ditto.
1. `aws eks update-kubeconfig` ditto.
1. `pinniped get-kubeconfig` ditto.

Reply:
> 2\. It seems weird to only remove that one CA data field from the `Cluster` struct. Maybe this is what you were already suggesting, but perhaps we should not set any of the `Cluster` info in the `KUBERNETES_EXEC_INFO` at all if someone doesn't turn on this flag?

yeah, I think opting into all the cluster info makes sense

> How would we indicate to the plugin (or to ourselves) that the Cluster data was not set in the env var?

Switch to this:
```go
  Cluster *Cluster `json:"cluster,omitempty"`
```

and it will be nil in the decoded ExecCredential object if it was omitted